src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -183,75 +194,104 @@ public boolean isCreated() {
-            String base64Model = Base64.getEncoder().encodeToString(model.getModelBlob());
-
-            Map<String, Object> parameters = ImmutableMap.of(
-                    KNNConstants.KNN_ENGINE, model.getModelMetadata().getKnnEngine().getName(),
-                    KNNConstants.METHOD_PARAMETER_SPACE_TYPE, model.getModelMetadata().getSpaceType().getValue(),
-                    KNNConstants.DIMENSION, model.getModelMetadata().getDimension(),
-                    KNNConstants.MODEL_BLOB_PARAMETER, base64Model
-            );
+            Map<String, Object> parameters = new HashMap<String, Object>() {{
+                put(KNNConstants.KNN_ENGINE, modelMetadata.getKnnEngine().getName());
+                put(KNNConstants.METHOD_PARAMETER_SPACE_TYPE, modelMetadata.getSpaceType().getValue());
+                put(KNNConstants.DIMENSION, modelMetadata.getDimension());
+                put(KNNConstants.MODEL_STATE, modelMetadata.getState().getName());
+                put(KNNConstants.MODEL_TIMESTAMP, modelMetadata.getTimestamp().toString());
+                put(KNNConstants.MODEL_DESCRIPTION, modelMetadata.getDescription());
+                put(KNNConstants.MODEL_ERROR, modelMetadata.getError());
+            }};
+
+            byte[] modelBlob = model.getModelBlob();
+
+            if (modelBlob == null && ModelState.CREATED.equals(modelMetadata.getState())) {
+                throw new IllegalArgumentException("Model blob cannot be null when model state is CREATED");
+            }
+
+            // Only add model if it is not null
+            if (modelBlob != null) {
+                String base64Model = Base64.getEncoder().encodeToString(modelBlob);
+                parameters.put(KNNConstants.MODEL_BLOB_PARAMETER, base64Model);
+            }
 
             IndexRequestBuilder indexRequestBuilder = client.prepareIndex(MODEL_INDEX_NAME, "_doc");
+
+            // Set id for request only if modelId is present
+            if (modelId != null) {
+                indexRequestBuilder.setId(modelId);
+            }
+
             indexRequestBuilder.setSource(parameters);
 
-            // Fail if the id already exists. Models are not updateable
-            indexRequestBuilder.setOpType(DocWriteRequest.OpType.CREATE);
+            indexRequestBuilder.setOpType(requestOpType); // Delegate whether this request can update based on opType
             indexRequestBuilder.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
 
-            // After the model is indexed, update metadata
-            ActionListener<IndexResponse> putMetadataListener = getUpdateModelMetadataListener(model.getModelMetadata(),
-                    listener);
+            // After metadata update finishes, remove item from cache if necessary. If no model id is
+            // passed then nothing needs to be removed from the cache
+            //TODO: Bug. Model needs to be removed from all nodes caches, not just local.
```

Comment:
Are we planning to fix this TODO in next PR? Having model updatable would result in these scenarios. We should Ideally be not updating them.

Reply:
Yes, this is a bug that needs to be fixed. The problem is more relevant in DELETE. #93

Reply:
Ok lets also put the issue in the comment.

Reply:
Will do.